[1.16 regression] determining exhaustive match on an enum with type members

Hi,

I continued looking at how the codebase I work on responds to the upcoming mypy 1.16 release and I believe I found another bug

```python
import datetime
import enum
from typing import assert_never


class ValueType(enum.Enum):
    BOOLEAN = bool
    DATE = datetime.date
    DATETIME = datetime.datetime


value_type: ValueType = ValueType.BOOLEAN

match value_type:
    case ValueType.BOOLEAN:
        pass
    case ValueType.DATE:
        pass
    case ValueType.DATETIME:
        pass
    case _:
        assert_never(value_type)
```

On mypy 1.16 I get this error

```
a.py:22: error: Argument 1 to "assert_never" has incompatible type "ValueType"; expected "Never"  [arg-type]
```

As far as I can tell this should be correct and I'm unsure why it doesn't think I have an exhaustive match.

**Your Environment**


- Mypy version used: release-1.16 (revision 96525a23f, built locally with mypy_mypyc-wheels for cp312-macosx_arm64)
- Mypy command-line flags: `mypy a.py`
- Mypy configuration options from `mypy.ini` (and other config files): no config
- Python version used: python3.12

I believe I found another example of this bug, and it seems to be more than just about enums specifically

So if we have two files

```python
# a.py
from typing import assert_never

import b


def switch[T_Choice: b.One | b.Two](choice: type[T_Choice]) -> None:
    match choice:
        case b.One:
            print(1)
        case b.Two:
            print(2)
        case _:
            assert_never(choice)


switch(b.One)

```

and

```python
# b.py
class One:
    pass


class Two:
    pass

```

Then I get no errors with mypy 1.15 and this error with 1.16

```
a.py:13: error: Argument 1 to "assert_never" has incompatible type "type[Two] | Never"; expected "Never"  [arg-type]
```

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -3301,8 +3301,8 @@ class TypeInfo(SymbolNode):
                         continue  # unannotated value not a member
 
                     typ = mypy.types.get_proper_type(sym.node.type)
-                    if isinstance(
-                        typ, mypy.types.FunctionLike
+                    if (
+                        isinstance(typ, mypy.types.FunctionLike) and not typ.is_type_obj()
                     ) or (  # explicit `@member` is required
                         isinstance(typ, mypy.types.Instance)
                         and typ.type.fullname == "enum.nonmember"

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -2638,3 +2638,24 @@ def f2() -> None:
                 return
     reveal_type(y) # N: Revealed type is "builtins.str"
 [builtins fixtures/list.pyi]
+
+[case testEnumTypeObjectMember]
+import enum
+from typing import NoReturn
+
+def assert_never(x: NoReturn) -> None: ...
+
+class ValueType(enum.Enum):
+    INT = int
+    STR = str
+
+value_type: ValueType = ValueType.INT
+
+match value_type:
+    case ValueType.INT:
+        pass
+    case ValueType.STR:
+        pass
+    case _:
+        assert_never(value_type)
+[builtins fixtures/tuple.pyi]

